OpenSim/Common/SmoothSegmentedFunction.cpp

```diff
@@ -39,6 +43,203 @@ static double UTOL = (double)SimTK::Eps*1e2;
+
+template<typename T>
+inline std::size_t HashCombine(std::size_t seed, T const &v) {
+    std::hash<T> hasher;
+    return seed ^ (hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2));
+}
+
+template<typename T>
+inline std::size_t HashOf(T const &v) {
+    return std::hash<T>{}(v);
+}
+
+template<typename T, typename... Others>
+inline std::size_t HashOf(T const &v, Others const &...others) {
+    return HashCombine(HashOf(v), HashOf(others...));
+}
+
+template<>
+inline std::size_t HashOf(SimTK::Matrix const &matrix) {
+    std::size_t hash = HashOf(matrix.nrow(), matrix.ncol());
+    for (int r = 0; r < matrix.nrow(); ++r) {
+        for (int c = 0; c < matrix.ncol(); ++c) {
+            hash = HashCombine(hash, matrix.row(r)[c]);
+        }
+    }
+    return hash;
+}
+
+} // namespace OpenSim
+
+template <>
+struct std::hash<SmoothSegmentedFunctionParameters> final
+{
+    size_t operator()(const SmoothSegmentedFunctionParameters& params) const {
+        return HashOf(
+            params._mX, params._mY,
+            params._x0, params._x1,
+            params._y0, params._y1,
+            params._dydx0, params._dydx1,
+            params._computeIntegral,
+            params._intx0x1
+        );
+    }
+};
+
+//=============================================================================
+// SMOOTHSEGMENTEDFUNCTION DATA
+//=============================================================================
+
+namespace OpenSim {
+struct SmoothSegmentedFunctionData {
+        SmoothSegmentedFunctionData(
+                const SmoothSegmentedFunctionParameters& params,
+                const std::string& name);
+
+        /**Array of spline fit functions X(u) for each Bezier elbow*/
+        SimTK::Array_<SimTK::Spline> _arraySplineUX;
+        /**Spline fit of the integral of the curve y(x)*/
+        SimTK::Spline _splineYintX;
+
+        /**Bezier X1,...,Xn control point locations. Control points are
+        stored in 6x1 vectors in the order above*/
+        SimTK::Array_<SimTK::Vector> _mXVec;
+        /**Bezier Y1,...,Yn control point locations. Control points are
+        stored in 6x1 vectors in the order above*/
+        SimTK::Array_<SimTK::Vector> _mYVec;
+
+        /**The number of quintic Bezier curves that describe the relation*/
+        int _numBezierSections;
+};
+
+//=============================================================================
+// DATA LOOKUP
+//=============================================================================
+
+/** Manages an unordered map of SmoothSegmentedFunction's Data, using the Parameters as key.
+ * If the same SmoothSegmentedFunctionParameters were previously used to
+ * construct the SmoothSegmentedFunctionData, a shared pointer to that data is
+ * obtained. If the given parameters are new, a new data object is constructed
+ * and added. This prevents duplication of the SmoothSegmentedFunction data for
+ * identical curves.
+*/
+class SmoothSegmentedFunctionDataCache final {
```

Comment:
This whole class should be defined in an anonymous namespace to prevent the ctor/assign/dtor/`lookup` from leaking into the public API of the resulting compilation unit (the class isn't used outside of this file?).